packages/amplify-provider-awscloudformation/src/utils/consolidate-apigw-policies.ts

```diff
@@ -0,0 +1,239 @@
+      expression: cdk.Fn.conditionEquals(env, 'NONE'),
+    });
+
+    let authRoleCount = 0;
+    let unauthRoleCount = 0;
+    let authPolicyDocSize = 0;
+    let unauthPolicyDocSize = 0;
+    let authManagedPolicy;
+    let unauthManagedPolicy;
+
+    props.apiGateways.forEach(apiGateway => {
+      if (!Array.isArray(apiGateway.params.paths)) {
+        return;
+      }
+
+      const apiRef = new cdk.CfnParameter(this, apiGateway.resourceName, {
+        type: 'String',
+      });
+
+      const state: ApiGatewayPolicyCreationState = {
+        apiGateway,
+        apiRef,
+        env,
+        path: null,
+        methods: null,
+        roleCount: 0,
+        roleName: null,
+        policyDocSize: 0,
+        managedPolicy: null,
+        namePrefix: '',
+      };
+
+      apiGateway.params.paths.forEach(path => {
+        state.path = path;
+
+        if (apiGateway.params.privacy.auth) {
+          state.methods = path?.privacy?.auth ?? [];
+          state.roleCount = authRoleCount;
+          state.roleName = authRoleName;
+          state.policyDocSize = authPolicyDocSize;
+          state.managedPolicy = authManagedPolicy;
+          state.namePrefix = 'PolicyAPIGWAuth';
+          this.createPoliciesFromResources(state);
+          ({ roleCount: authRoleCount, policyDocSize: authPolicyDocSize, managedPolicy: authManagedPolicy } = state);
+        }
+
+        if (apiGateway.params.privacy.unauth) {
+          state.methods = path?.privacy?.unauth ?? [];
+          state.roleCount = unauthRoleCount;
+          state.roleName = unauthRoleName;
+          state.policyDocSize = unauthPolicyDocSize;
+          state.managedPolicy = unauthManagedPolicy;
+          state.namePrefix = 'PolicyAPIGWUnauth';
+          this.createPoliciesFromResources(state);
+          ({ roleCount: unauthRoleCount, policyDocSize: unauthPolicyDocSize, managedPolicy: unauthManagedPolicy } = state);
+        }
+      });
+    });
+  }
+
+  toCloudFormation() {
+    prepareApp(this);
+    return this._toCloudFormation();
+  }
+
+  private createPoliciesFromResources(options: ApiGatewayPolicyCreationState) {
+    const { apiGateway, apiRef, env, roleName, path, methods, namePrefix } = options;
+
+    methods.forEach(method => {
+      const policySizeIncrease = computePolicySizeIncrease(method.length, path.policyResourceName.length, apiGateway.resourceName.length);
+
+      options.policyDocSize += policySizeIncrease;
+      // If a managed policy hasn't been created yet, or the maximum
+      // managed policy size has been exceeded, then create a new policy.
+      if (options.roleCount === 0 || options.policyDocSize > MAX_MANAGED_POLICY_SIZE) {
+        // Initial size of 100 for version, statement, etc.
+        options.policyDocSize = 100 + policySizeIncrease;
+        options.roleCount++;
+        options.managedPolicy = createManagedPolicy(this, `${namePrefix}${options.roleCount}`, (roleName as unknown) as string);
```

Comment:
won't this overwrite the previous policy on each iteration?

Reply:
Only when a new policy needs to be created (either because it's the first iteration, or because the current policy is "full").